core/domain/user_id_migration.py

```diff
@@ -0,0 +1,518 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Jobs for queries personalized to individual users."""
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import inspect
+
+from core import jobs
+from core.platform import models
+import feconf
+
+(
+    base_models, collection_models,
+    exp_models, question_models, skill_models,
+    topic_models, user_models) = models.Registry.import_models(
+        [models.NAMES.base_model, models.NAMES.collection,
+         models.NAMES.exploration, models.NAMES.question, models.NAMES.skill,
+         models.NAMES.topic, models.NAMES.user])
+datastore_services = models.Registry.import_datastore_services()
+transaction_services = models.Registry.import_transaction_services()
+
+
+class MissingUserException(Exception):
+    """Exception for cases when the user doesn't exist."""
+    pass
+
+
+class UserIdMigrationJob(jobs.BaseMapReduceOneOffJobManager):
+    """One-off job for creating new user ids for all the users and re-adding
+    models that use the user id. This migration doesn't handle snapshot content
+    models that can contain user ID, these are handled by
+    SnapshotsUserIdMigrationJob.
+    """
+
+    @staticmethod
+    def _copy_model_with_new_id(model_class, old_user_id, new_user_id):
+        """Create new model with same values but new id.
+
+        Args:
+            model_class: class. The class of the migrated model.
+            old_user_id: str. The old (GAE) ID of the user being migrated.
+            new_user_id: str. The newly generated ID of the user being migrated.
+
+        Returns:
+            optional((str, (str, str)).
+        """
+        old_model = model_class.get_by_id(old_user_id)
+        if not old_model:
+            # Some models are defined only for some users (for example
+            # UserSubscribersModel, is only defined for users who actually have
+            # at least one subscriber) that is why we are okay with the fact
+            # that model is None.
+            return ('MISSING OLD MODEL', (old_user_id, new_user_id))
+        model_values = old_model.to_dict()
+        model_values['id'] = new_user_id
+        new_model = model_class(**model_values)
+
+        def _replace_model():
+            """Replace old model with new one."""
+            new_model.put(update_last_updated_time=False)
+            old_model.delete()
```

Comment:
Just a note -- in some models (e.g. subclasses of VersionedModel) I believe deletion is currently implemented softly (i.e. setting deleted to True) rather than deleting the model. Could you please double-check that any models that are deleted by this PR are actually fully deleted (and don't exist any more), and add tests for that?

Reply:
I checked all the models with `USER_ID_MIGRATION_POLICY` set to `COPY` or `COPY_AND_UPDATE_ONE_FIELD`, all these models have the `delete` method implemented as force deletion by default (none of these models inherits from `VersionedModel`). Also, in the tests, I use `get_by_id` (not affected by our `deleted` param) to check that the old version of the model was hard deleted.